### PR TITLE
[TF circular reference] Allow specifying mutually dependant components in PCL

### LINKED ISSUE #17760: [TF circular references] Allow specifying mutually dependant components in PCL
### Description

First part required to address https://github.com/pulumi/pulumi/issues/13581 is allowing the PCL binder to _not_ error out when encountering mutually dependant components

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
Fixes #17760 by changing the condition to error out when encountering circular references. For PCL components we don't error with a circular ref when the components
 - are not self-referencing
 - have mutual dependencies with other components

Added unit tests to assert the new behavior 

Breaking the circular reference will be a responsibility of each of the target languages we support.

### CODE REVIEW

File: pkg/codegen/pcl/binder_test.go [hunk 1]
```diff
@@ -749,3 +749,84 @@ value = element(data, 0)
 	assert.Empty(t, diags, "There are no error or warning diagnostics")
 	assert.NotNil(t, program)
 }
+
+func TestBindingSelfReferencingComponentFailsWithCircularReference(t *testing.T) {
```
Comment (brandonpollack23) [inline pkg/codegen/pcl/binder_test.go:753]: nit: WithSelfReferenceWhenCircular

or something to that effect?

File: pkg/codegen/pcl/binder_nodes.go [hunk 2]
```diff
@@ -22,20 +22,57 @@ import (
...
+		// - the circular reference is only between other nodes of type Component
+		if !isComponent(node) || selfReferencingNode(node) {
+			// TODO(pdg): print better trace
+			rng := node.SyntaxNode().Range()
+			return hcl.Diagnostics{{
+				Severity: hcl.DiagError,
+				Summary:  "circular reference",
+				Subject:  &rng,
+			}}
+		}
```
Comment (brandonpollack23) [inline pkg/codegen/pcl/binder_nodes.go:75]: else should we have a warning or an info diagnostic of some sort?

**Review by brandonpollack23 — APPROVED** [review]

File: pkg/codegen/pcl/binder_nodes.go (hunk 2 above)
Comment (Frassle) [reply pkg/codegen/pcl/binder_nodes.go:75]: No, this isn't a warning case. If it's a circular dependency between components that will just work once we've fixed codegen up to handle it.

File: pkg/codegen/pcl/binder_test.go [hunk 3]
```diff
@@ -749,3 +749,84 @@ value = element(data, 0)
...
+		pcl.DirPath(absoluteProgramPath),
+		pcl.ComponentBinder(pcl.ComponentProgramBinderFromFileSystem()))
+
+	assert.Nil(t, program)
+	assert.NotNil(t, bindError)
+	assert.True(t, diags.HasErrors(), "There are error diagnostics")
+	assert.Contains(t, diags.Error(), "circular reference")
+}
+
+func TestBindingMutuallyDependantComponentsSucceeds(t *testing.T) {
```
Comment (Frassle) [inline pkg/codegen/pcl/binder_test.go:794]: Do we have a test for the current error that circular deps between resources results in error? If not that needs adding.

Comment (Zaid-Ajaj) [reply pkg/codegen/pcl/binder_test.go:794]: I'll check and add one if we don't

File: pkg/codegen/pcl/binder_test.go (hunk 1 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/pcl/binder_test.go:753]: "SelfReferencing" already implies a circular reference with self :thinking: I read the test like this:
"Test binding self-referencing components fails with a circular reference error" 

I can add a comment about the test if it is not clear enough

File: pkg/codegen/pcl/binder_test.go (hunk 3 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/pcl/binder_test.go:826]: Done ✅

File: pkg/codegen/pcl/binder_test.go (hunk 1 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/pcl/binder_test.go]: Added a comment explaining what the test does ✅

**Pull request merged**